src/dune_console/dune_console.ml

```diff
@@ -131,26 +143,72 @@ module Backend = struct
 
   let spawn_thread = Fdecl.create Dyn.opaque
 
-  let threaded (module Base : S) : (module S) =
+  (** Backends may have access to an internal state. This type is used by the
+      [Threaded] backends. *)
+  type state =
+    { messages : User_message.t Queue.t
+    ; mutable finish_requested : bool
+    ; mutable finished : bool
+    ; mutable status_line : User_message.Style.t Pp.t option
+    }
+
+  (** [Threaded] is the interface for user interfaces that are rendered in a
+      separate thread. It can be used to directly construct a threaded backend.
+
+      This module interface will be used by both the current console UI and the
+      upcoming terminal UI (TUI).
+
+      Notice that the TUI can also react to keyboard input, so it has the
+      additional [handle_user_events] call. *)
+  module type Threaded = sig
+    (** [start] is called by the main thread to start broadcasting the user
+        interface. Any initial setup should be performed here. *)
+    val start : unit -> unit
+
+    (** [render state] is called by the main thread to render the current state
+        of the user interface. *)
+    val render : state -> unit
+
+    (** [handle_user_events ~now ~timeout mutex] is called by the main thread to
+        handle user events such as keypresses. The function should return the
+        time at which the next event should be handled. A [mutex] is provided in
+        order to lock the state of the UI.*)
+    val handle_user_events : now:float -> timeout:float -> Mutex.t -> float
```

Comment:
Can you document what `timeout` does?

Reply:
Timeout tells handle_user_events an amount of time to wait. I'll add some doc.

Reply:
handle_user_events may take some time to do, so knowing how long it was supposed to take can let it calculate how long it should really take.

Reply:
In other words, timeout is the maximum amount of time handle_user_events should take. This is an important piece of information for other implementations of handle_user_events that listen for user inputs, since they will need to wait in order to do that.

Reply:
I see, thanks! Might we call it `time_budget` or something like this instead? I find `timeout` a little ambiguous.

Reply:
OK, I've renamed.